src/kepler/ipc.zig

```diff
@@ -0,0 +1,582 @@
+        RequestPending,
+        /// Sent to client's notification queue when its request
+        /// to make a stream was accepted. Once recieved, kernel
+        /// performs the work needed for it to be used to exchange
+        /// further notifications
+        RequestAccepted,
+        /// Sent to client's notification queue when its request
+        /// to make a stream was denied. Once recieved, kernel
+        /// transfers consumer to the .Denied state, in which
+        /// the only valid call on the stream would be close()
+        RequestDenied,
+        /// Sent to server's notification queue with the intent
+        /// to notify the server that more tasks are available
+        /// for it to handle.
+        TasksAvailable,
+        /// Sent to client's notification queue with the intent
+        /// to notify the client that some of the task were
+        /// completed and their results are available to the client
+        ResultsAvailable,
+        /// Sent to the server's notification queue when client leaves
+        /// to notify the server that client has abandoned the stream
+        ConsumerLeft,
+        /// Sent to the client's notification queue when server leaves
+        /// to notify the client that server has abandoned the stream
+        ProducerLeft,
+        /// Sent to the server's notification queue to indicate that
+        /// all endpoints are lost. The intent is to allow server
+        /// to cleanup all the structures that were used for the endpoint
+        EndpointUnreachable,
+        /// Sent to the driver whenever owned InterruptObject
+        /// raises interrupt
+        InterruptRaised,
+
+        /// Get peer type from the note type (whether this note is directed
+        /// to producer or consumer)
+        pub fn to_peer_type(self: @This()) Stream.Peer {
+            return switch (self) {
+                .RequestPending, .TasksAvailable, .ConsumerLeft, .EndpointUnreachable => .Producer,
+                .RequestAccepted, .RequestDenied, .ResultsAvailable, .ProducerLeft => .Consumer,
+                .InterruptRaised => @panic("to_peer_type called on a wrong message type"),
+            };
+        }
+    };
+
+    /// Hook for the atomic queue in NoteQueue
+    hook: atmcqueue.Node = undefined,
+    /// Type of the notification
+    typ: Type,
+    /// Borrowed reference to the owner
+    owner_ref: union {
+        /// For all the note types except for .EndpointUnreachable, this
+        /// field stores reference to the stream object over which message was sent
+        stream: *Stream,
+        /// For the .EndpointUnreachable, this field stores the reference to the endpoint,
+        /// all references to which are already lost
+        endpoint: *Endpoint,
+        /// For the .interruptRaised, this field stores the reference to the InterruptObject
+        /// that has raised the interrupt
+        interrupt: *kepler.interrupts.InterruptObject,
+    },
```

Comment:
Should this use the `Type` enum since the contents depend on the `Type`, and then the "enum active member" and the `typ` field can be merged into one?
```
owner_ref: union(Type) {
  .EndpointUnreachable: *Stream,
  // ...
};
```